lib/internal/modules/cjs/loader.js

```diff
@@ -1015,6 +1029,153 @@ function getExportsForCircularRequire(module) {
+
+  const resolveResult = resolveWithHooks(specifier, parentURL, /* importAttributes */ undefined,
+                                         getCjsConditions(), defaultResolve);
+  const { url } = resolveResult;
+  format = resolveResult.format;
+
+  let filename;
+  if (url === defaultResolvedURL) {  // Not overridden, skip the re-conversion.
+    filename = defaultResolvedFilename;
+  } else {
+    filename = convertURLToCJSFilename(url);
+  }
+
+  return { __proto__: null, url, format, filename, parentURL };
+}
+
+/**
+ * @typedef {import('internal/modules/sync_hooks').ModuleLoadContext} ModuleLoadContext;
+ * @typedef {import('internal/modules/sync_hooks').ModuleLoadResult} ModuleLoadResult;
+ */
+
+/**
+ * Load the source code of a module based on format.
+ * @param {string} filename Filename of the module.
+ * @param {string|undefined|null} format Format of the module.
+ * @returns {string|null}
+ */
+function defaultLoadImpl(filename, format) {
+  switch (format) {
+    case undefined:
+    case null:
+    case 'module':
+    case 'commonjs':
+    case 'json':
+    case 'module-typescript':
+    case 'commonjs-typescript':
+    case 'typescript': {
+      return fs.readFileSync(filename, 'utf8');
+    }
+    case 'builtin':
+      return null;
+    default:
+      // URL is not necessarily necessary/available - convert it on the spot for errors.
+      throw new ERR_UNKNOWN_MODULE_FORMAT(format, convertCJSFilenameToURL(filename));
+  }
+}
+
+/**
+ * Construct a last nextLoad() for load hooks invoked for the CJS loader.
+ * @param {string} url URL passed from the hook.
+ * @param {string} filename Filename inferred from the URL.
+ * @returns {(url: string, context: ModuleLoadContext) => ModuleLoadResult}
+ */
+function getDefaultLoad(url, filename) {
+  return function defaultLoad(urlFromHook, context) {
+    // If the url is the same as the original one, save the conversion.
+    const isLoadingOriginalModule = (urlFromHook === url);
+    const filenameFromHook = isLoadingOriginalModule ? filename : convertURLToCJSFilename(url);
+    const source = defaultLoadImpl(filenameFromHook, context.format);
+    // Format from context is directly returned, because format detection should only be
+    // done after the entire load chain is completed.
+    return { source, format: context.format };
+  };
+}
+
+/**
+ * Load a specified builtin module, invoking load hooks if necessary.
+ * @param {string} id The module ID (without the node: prefix)
+ * @param {string} url  The module URL (with the node: prefix)
+ * @param {string} format Format from resolution.
+ * @returns {any} If there are no load hooks or the load hooks do not override the format of the
+ *                builtin, load and return the exports of the builtin. Otherwise, return undefined.
+ */
+function loadBuiltinWithHooks(id, url, format) {
+  if (loadHooks.length) {
+    url ??= `node:${id}`;
+    // TODO(joyeecheung): do we really want to invoke the load hook for the builtins?
```

Comment:
Maybe the result could be asserted that `loadResult.source === null` if the format is not overridden, so that the result is not silently discarded?

Reply:
Sounds good to me